Python metaclasses for metaprogramming (#5730)
section-engineering-education/engineering-education: content/articles/python-metaclasses-for-metaprogramming/index.md

---
layout: engineering-education
status: publish
published: true
url: /python-metaclasses-for-metaprogramming/
title: Understanding Metaclasses in Python
description: This tutorial discusses how to use python's metaclasses for metaprogramming. This process involves writing code that manipulates other code. 
author: dickson-gitau
date: 2021-12-23T00:00:00-14:10
topics: [Languages]
excerpt_separator: <!--more-->
images:

  - url: /engineering-education/python-metaclasses-for-metaprogramming/hero.jpg
    alt: Python Metaclasses for Metaprogramming Hero Image
---

> Metaprogramming is the practice of writing code that manipulates other code. It refers to the ability of a program to regulate itself.

Reviewer: It would be really great if you expounded on how a program regulates itself.

<!--more-->
In Python,  metaprogramming can be implemented using decorators and metaclasses.

> Object-oriented programming(OOP) is essential in Python. This concept relies on metaprogramming. 

Reviewer: Form a complete paragraph from these sentences. It is unprofessional to have one line paragraphs

> Object-oriented programming(OOP) is essential in Python. This concept relies on metaprogramming. 

Reviewer: Tell the readers how OOP relies on metaprogramming.

### Prerequisites

> To follow along, you need some some knowledge of decorators and object-orented programming concepts in Python.

Reviewer: ```suggestion
To follow along with this article, you need to have:
- Knowledge of decorators 
- Object-oriented programming concepts in Python.

```

### Table of contents
- [What are metaclasses](#what-are-metaclasses)
- [Understanding metaclasses with examples](#understanding-metaclasses-with-examples)
- [Reasons to use metaclasses instead of functions](#reasons-to-use-metaclasses-instead-of-functions)
- [Instances to use metaclasses](#instances-to-use-metaclasses)
- [Customizing metaclasses](#customizing-metaclasses)
- [Using a metaclass to fix an issue](#using-a-metaclass-to-fix-an-issue)
- [Conclusion](#conclusion)

### What is a metaclass?

> Metaclasses in Python are classes of classes that specify the behavior of a class. A class is a metaclass in its own right. How the class is implemented is defined by its class definition in Python.

Reviewer: This point is not clear. I am having trouble understanding. Would you please elaborate?

> We need to add additional code or processes to the Metaclass framework. Although end-users seldom use Metaclasses, they are essential in some programming problems.

Reviewer: Explain why we add additional code. What are these processes?

```Python
class Town:  
    pass  
Town.b = 70
Town.foo = lambda persona: print('Dusk')
enjoyerobject = Town()  
   
print(enjoyerobject.b)  
enjoyerobject.foo() 
```

#### Output

```Python
70
Dusk
```

We may proceed by starting with a class that does not utilize any methods or variables from the class. Then, the method variables and class methods will be specified in the following steps. In the end, we will create the final object.

### Reasons to use metaclasses instead of functions
Although we have equally useful functions, we need metaclasses for the following reasons.
- The aim of metaclasses is precise. At the point when you read `UpperAttrMetaclass(type)`, you know what will follow
- You can utilize OOP. Metaclasses may inherit the parent method.
- A class's subclasses are instances of its metaclass if the metaclass-class is given, but not the metaclass function.
- You may improve the organization of your code. Using metaclasses for a simple example like this is a waste of time. It is usually reserved for a more difficult task. Making several methods and grouping them into a single class makes the code more readable.

> - It is possible to hook on `new`, `init`, and `call` hooks. 

Reviewer: @Dick7603  These points are great but you need to clarify them. Provide some in-depth explanation for the benefit of readers.

### Instances to use metaclasses
The developers seldom utilizes metaclasses since they are primarily employed in complex scenarios. However, metaclasses may be used in the following limited number of situations:
- While generating down the hierarchies for an inheritance, you can employ a metaclass. This will have an impact on all subclasses as well. If the user finds themselves in such a scenario, they can use the metaclass.
- To change the class automatically, they can utilize a metaclass when it is made.
- Assuming the user is an application programming interface designer, they can utilize metaclasses for that reason.

> 

Reviewer: Explain these points well and provide real-life examples.
